### PR TITLE
proposed fix for #707

### PR DESCRIPTION
**Pull request by gavinking**
Make sure that queued requests use the right `session` and `entity` args.

### CODE REVIEW

**Comment by gavinking** [comment]
OK, @Sanne, @DavideD, @stuartwdouglas, WDYT of this?

Obviously I have not tested it.

**Comment by Sanne** [comment]
I think we need good tests for this area, not least to check how this behaves when the consumer rate is higher than our ability to produce new ids.

**Comment by gavinking** [comment]
> I think we need good tests for this area

Well, yeah, clearly. The problem is that we have to precisely no tests for concurrent operation in the HR project, and I'm not clear even on the right way to write them.

**Comment by gavinking** [comment]
(I force-pushed to remove the useless `Queued` class and just use a `Runnable` instead.)

**Comment by stuartwdouglas** [comment]
Looks good, but if we have n consumers waiting maybe we should make sure we request at least n IDs? At the moment by default it is one at a time, so this is basically a global lock in terms of persist performance.

**Comment by gavinking** [comment]
> Looks good, but if we have n consumers waiting maybe we should make sure we request at least n IDs? At the moment by default it is one at a time, so this is basically a global lock in terms of persist performance.

Well the assumption I made was that you always choose a big enough block size that you "rarely" need to request a new block. If that's true, then I think having more requests waiting than the size of the block is an extremely low-probability event.

On the other hand, what I realized earlier was that, ooops, what about people who *don't* want blocks, and set the block size to 1? We should actually be bypassing all this logic entirely in that case, and we don't. Not quite sure how I overlooked that.

**Comment by gavinking** [comment]
> if we have n consumers waiting maybe we should make sure we request at least n IDs?

Also, at the time we request an hi value, we have zero consumers waiting.

**Comment by gavinking** [comment]
> Not quite sure how I overlooked that.

Well, hrm, actually I supposed it's *not* entirely clear what the best behavior here is.

For *sequences*, yes, I think the right thing to do is not *not* wait for concurrent requests to complete.

For *table* generation it's not so clear, however.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/id/impl/BlockingIdentifierGenerator.java [hunk 1]
```diff
@@ -54,6 +54,11 @@ protected synchronized long next(long hi) {
 
     @Override
     public CompletionStage<Long> generate(ReactiveConnectionSupplier session, Object entity) {
+        if ( getBlockSize()<=1 ) {
+            //special case where we're not using blocking at all
+            return nextHiValue(session);
+        }
+
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/id/impl/BlockingIdentifierGenerator.java:61]: @stuartwdouglas WDYT?

**Comment by Sanne** [comment]
FWIW I had written a POC about a year ago to have such an optimizer, which would tune the page sized in an "adaptative" way.

I abandoned it as I needed to learn how to test such things with vertx and never had much time - but it sounds like conceptually you might want to revive the idea.

**Comment by gavinking** [comment]
I mean in principle it would not be at all hard to increase the block size if we find ourselves requesting too many new blocks. Not hard at all.

**Comment by Sanne** [comment]
Right - and there's several other smart things one can easily do in the reactive model, such as we could already trigger a block allocation before the last available ID has been handed out to clients.

So by observing the rate at which IDs are being consumed and the time it takes for block allocation tasks to complete it should be possible to:

- prefetch ID blocks in advance
- adapt the block sizes

One caveat of using dynamically allocated block sizes might be fragmentation; if there's multiple application nodes they still need to use a consistent strategy - should not be a problem when it's based on sequences (atomic) but if it's backed by some of the other models (table?) this is probably not a good idea.

**Comment by gavinking** [comment]
@stuartwdouglas can you do us a favor and test this patch in the context of the Quarkus issue?

**Comment by DavideD** [comment]
> > I think we need good tests for this area
> 
> Well, yeah, clearly. The problem is that we have to precisely no tests for concurrent operation in the HR project, and I'm not clear even on the right way to write them.

I think I have an idea, I will start prototyping something

**Comment by stuartwdouglas** [comment]
This works now, instead of doing a couple of requests then hanging it can now handle 3.5k inserts per second (this is not tuned in any way, it could probably do a lot more if I messed with some settings).

**Comment by gavinking** [comment]
Thanks 🙏 

So perhaps it's worth doing a release just to fix this one thing. Thoughts?

**Comment by stuartwdouglas** [comment]
I think so, its a pretty bad bug.

**Comment by stuartwdouglas** [comment]
Actually I just realized that I am only testing the  block size == 1 path, how do I configure a larger block size to test the other code path?

**Comment by gavinking** [comment]
With the `@SequenceGenerator` or `@TableGenerator` annotation just like in regular JPA.

**Comment by DavideD** [comment]
> So perhaps it's worth doing a release just to fix this one thing. Thoughts?

Yes, we should release it. I will work on adding some more tests for the id generation but considering that @stuartwdouglas already tested on quarkus and that our current test suite works, releasing this version seems an important improvement for a release.

**Comment by stuartwdouglas** [comment]
It has been more than 10 years since I last actually 'used' JPA rather than just dealing with integration stuff. So it looks like the quickstart sets allocationSize = 1 for some reason, if you remove it then performance increases, but we get the occasional PK violation:

```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.86ms   17.38ms 219.92ms   98.11%
    Req/Sec     8.11k     1.49k    9.41k    90.86%
  159144 requests in 10.10s, 15.69MB read
  Non-2xx or 3xx responses: 31

```

Looking at the logs some entities are being given the same PK.

**Comment by gavinking** [comment]
> we get the occasional PK violation

is this with a sequence generator or a table generator?

**Comment by stuartwdouglas** [comment]
```
    @SequenceGenerator(name = "fruitsSequence", sequenceName = "known_fruits_id_seq", initialValue = 10)
```

**Comment by gavinking** [comment]
Well that's very strange. Can you tell if it's because they are assigned dupe hi values or dupe lo values?

**Comment by gavinking** [comment]
I mean: 

1. a dupe hi value is impossible because we're getting them from a database sequence, but
2. a dupe lo value also looks impossible because lo values are assigned within a `synchronized` block.

So ... another silly logical error in my code??

**Comment by stuartwdouglas** [comment]
The id's in the errors are a long way apart, e.g:
5161
6361
8561
13211

So it's not allocating duplicate blocks.

**Comment by stuartwdouglas** [comment]
They all seem to end in '1'

**Comment by gavinking** [comment]
> another silly logical error in my code

Yup it's that, actually. Hold tight I will push a fix.

**Comment by gavinking** [comment]
wait, *all* the ids, or just all the *dupe* ids?

**Comment by gavinking** [comment]
> Yup it's that, actually. Hold tight I will push a fix.

No, excuse me. What I thought was wrong is actually fine.

**Comment by gavinking** [comment]
> wait, all the ids, or just all the dupe ids?

No, you mean the ones in the errors. So by the looks, they're all the "second id generated after fetching a new hi value".

**Comment by gavinking** [comment]
> they're all the "second id generated after fetching a new hi value".

No, wait I guess you have `initialValue=1`, the default, so they're the *first* id's in their respective blocks.

**Comment by stuartwdouglas** [comment]
I had to step away and have dinner, but my guess is something to do with
reentrant calls, so the sync block doesn't help. I can't see how though.

**Comment by gavinking** [comment]
Do you think it could be an out-of-order execution thing??

Like, `list.forEach(Runnable::run)` gets executed before ` result.complete( next(id) )`.

Or am I trying to make this too complicated again?

**Comment by gavinking** [comment]
I mean, would moving the line `list.forEach(Runnable::run)` inside the `synchronized` block fix it?

**Comment by stuartwdouglas** [comment]
Bugger, turns out this was user error. I forgot to clean before running quarkus:dev so it had cached the old maven dependencies, so it was not picking up the new version. It looks like this was a problem with the old version when the block size was larger than 1.

Sorry for the noise.

**Comment by gavinking** [comment]
No problem. So how does it perform?

**Comment by stuartwdouglas** [comment]
Comparatively way better, 13,429 inserts per sec. This is no no way a proper benchmark, it's only using 2 threads and 10 connections.

**Comment by gavinking** [comment]
Alright. So that sounds good.

Shall I merge this?

**Comment by DavideD** [comment]
Sounds good to me

**Comment by DavideD** [comment]
I've released it and sent a PR for quarkus: https://github.com/quarkusio/quarkus/pull/16588

**Comment by DavideD** [comment]
Thanks a lot @stuartwdouglas

**Pull request merged**